### ISSUE TITLE
Fake overlap callbacks for FlxSprite after they no longer overlap

### ISSUE THREAD

**Issue by starry-abyss** [HaxeFlixel/flixel#2059]
- **Haxe version:** 3.2.0
- **Flixel version:** cddeaae
- **OpenFL version:** 3.6.1
- **Lime version:** 2.9.1
- **Affected targets:** neko, flash

________________________________

And also two "good overlap"'s and two "fake overlap"'s are triggered in the same frame (two overlaps per frame).

**Code snippet reproducing the issue**:

```haxe
package;

import flixel.FlxG;
import flixel.FlxSprite;
import flixel.FlxState;
import flixel.input.mouse.FlxMouseEventManager;
import flixel.math.FlxVector;
import flixel.text.FlxText;
import flixel.tile.FlxTilemap;
import flixel.ui.FlxButton;
import flixel.math.FlxMath;
import flixel.util.FlxColor;

class PlayState extends FlxState
{
	var enemySprite: FlxSprite;
	var playerSprite: FlxSprite;
	
	override public function create():Void
	{
		super.create();

		enemySprite = new FlxSprite(100, 300);
		enemySprite.makeGraphic(100, 100, FlxColor.RED);
		add(enemySprite);
		
		playerSprite = new FlxSprite(100, 100);
		playerSprite.makeGraphic(100, 100, FlxColor.GREEN);
		add(playerSprite);
		
		playerSprite.velocity.y = 100;
	}

	override public function update(elapsed:Float):Void
	{
		super.update(elapsed);

		FlxG.overlap(playerSprite, enemySprite, enemyCallback);
	}
	
	function enemyCallback(playerSprite: FlxSprite, enemySprite: FlxSprite): Void
	{
		if (!playerSprite.overlaps(enemySprite))
		{
			trace("fake overlap");
			return;
		}
		
		trace("good overlap");
		
		playerSprite.velocity.y = -100;
	}
}
```

________________________________

**Observed behavior:**
good overlap
good overlap
fake overlap
fake overlap

**Expected behavior:**
good overlap

**Comment by MSGhero** [comment]
The multiple overlap thing is #1247, unsure of how overlaps work re: the second part.

**Comment by Gama11** [comment]
As @MSGhero said, this seems like a duplicate of #1247.

**Comment by starry-abyss** [comment]
No, I don't think it is. Adding `FlxG.worldDivisions = 1` gives:

```
good overlap
fake overlap
```

which is without multiple overlaps but still with 'fake' ones.

**Comment by danim1130** [comment]
I wasn't able to test this, but looking at the source code, this is caused by the inconsistance of the overlaps functions.

See, FlxG.overlap() is used by FlxG.collide(). To make sure that fast objects like bullets don't go through small objects, FlxG.overlap() compares with a bigger rectangle that is composite of the object's "hit-box" at the last and the current frame. In turn, FlxObject.overlaps() only looks at the current frame, so the hit-box becomes smaller.

Basically, if you want to look for overlap as "is my object currently on another object", use FlxObject.overlaps(). If you want to look for "has my object overlapped with the other object during their movement from the last frame", use FlxG.overlap().

Maybe this should be added to the documentation?

**Comment by starry-abyss** [comment]
Thanks, but it means that I can't do group overlap check only for current frame? Only either call .overlaps() object by object or check the same thing 3 times (overlap current, overlap last, overlaps current) like in the example?

**Comment by danim1130** [comment]
Unfortunately, that seems about right, though it's been a while since I
worked on flixel.

As a workaround, you probably don't need the position of the last frame (if
you don't do anything related to collisions). In that case, (if there are
lot of objects in the groups, it might be better to call reset() on every
object in the groups (which should reset the last frame positions), then
you can safely call flxg.overlap(). Note that you will have to call reset
any time the position changes if you want to go down this route.

**Comment by Gama11** [comment]
Since it behaves this way by design, as @danim1130 explained, I think this can be closed.
